commit c72f27527f5a7b070c14a34c1ba31a59bfd9e59e (refs/changes/32/2432/10)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-09-09T00:49:36+00:00 (2 years, 1 month ago)
    
    MANTA-3220 MPU shark selection should try to connect to several sets of sharks

diff --git a/lib/uploads/create.js b/lib/uploads/create.js
index 5b35117..6f69aa7 100644
--- a/lib/uploads/create.js
+++ b/lib/uploads/create.js
@@ -16,6 +16,7 @@ var path = require('path');
 var restify = require('restify');
 var util = require('util');
 var verror = require('verror');
+var vasync = require('vasync');
 
 var auth = require('../auth');
 var common = require('../common');
@@ -32,6 +33,78 @@ var sprintf = util.format;
 
 ///--- Helpers
 
+
+/*
+ * Attempts to grab a running pool for a given shark from cueball
+ * to ensure that muskie had a recent active connection to the
+ * shark. If it didn't, one will be created with the cueball
+ * createPool api method. This function is used in chooseSharks to
+ * pick an appropriate shark for a new mpu-create operation.
+ */
+function ensureSharkHasPool(req, attempted, shark, callback) {
+    var host = shark.manta_storage_id;
+    var pool = req.sharkAgent.getPool(host);
+    if (pool) {
+        req.log.debug({
+            shark: shark,
+            pool: pool.getState()
+        }, 'ensureSharkHasPool: got existing pool');
+    }
+
+    if (!pool) {
+        req.sharkAgent.createPool(shark.manta_storage_id);
+        pool = req.sharkAgent.getPool(host);
+        req.log.debug({
+            shark: shark,
+            pool: pool.getState()
+        }, 'ensureSharkHasPool: created pool for shark');
+    }
+
+    if (pool.isInState('running')) {
+        setImmediate(callback, null, shark);
+    } else if (pool.isInState('starting')) {
+        var timedOut = false;
+
+        var sharkTimeout = setTimeout(function () {
+            timedOut = true;
+            req.log.debug({
+                shark: shark,
+                state: pool.getState()
+            }, 'ensureSharkHasPool: pool state change timeout');
+            attempted.push(shark);
+            callback(new Error('invalid state'));
+        }, req.sharkConfig.connectTimeout);
+
+        function onStateChanged(st) {
+            if (timedOut) {
+                return;
+            }
+            clearTimeout(sharkTimeout);
+            req.log.debug({
+                shark: shark,
+                state: st
+            }, 'ensureSharkHasPool: pool state changed');
+            if (st === 'running') {
+                callback(null, shark);
+            } else if (st === 'starting') {
+                pool.once('stateChanged', onStateChanged);
+            } else if (st === 'failed') {
+                attempted.push(shark);
+                callback(new Error('shark pool failed'));
+            }
+        }
+        pool.once('stateChanged', onStateChanged);
+
+    } else {
+        req.log.debug({
+            shark: shark,
+            pool: pool.getState()
+        }, 'ensureSharkHasPool: pool neither running nor starting');
+        setImmediate(callback, null, new Error('invalid pool state'));
+    }
+}
+
+
 /*
  * Selects the sharks for the upload through the picker.choose interface.
  *
@@ -39,7 +112,7 @@ var sprintf = util.format;
  * the durability-level and the content-length headers, respectively, or
  * set to a default value.
  */
-function chooseSharks(req, size, copies, cb) {
+function chooseSharks(req, res, size, copies, cb) {
     assert.object(req, 'req');
     assert.number(size, 'size');
     assert.number(copies, 'copies');
@@ -55,14 +128,37 @@ function chooseSharks(req, size, copies, cb) {
             replicas: copies,
             size: size
         };
+
         req.picker.choose(opts, function (err, sharks) {
             if (err) {
                 cb(err);
             } else {
-                log.debug({
-                    sharks: sharks[0]
-                }, 'upload: sharks chosen');
-                cb(null, sharks[0]);
+                var ndx = 0;
+                var attempted = [];
+
+                function attempt(inputs) {
+                    vasync.forEachParallel({
+                        func: ensureSharkHasPool.bind(null, req, attempted),
+                        inputs: inputs
+                    }, function (err2, results) {
+                        if (err2 || results.successes.length < copies) {
+                            if (ndx < sharks.length) {
+                                attempt(sharks[ndx++]);
+                            } else {
+                                cb(new SharksExhaustedError(res), null);
+                            }
+                            return;
+                        }
+                        log.info({
+                            sharks: attempted
+                        }, 'chooseSharks: unsuccessfully attempted sharks');
+                        log.info({
+                            sharks: results.successes
+                        }, 'chooseSharks: sharks chosen');
+                        cb(null, results.successes);
+                    });
+                }
+                attempt(sharks[ndx++]);
             }
         });
     }
@@ -285,7 +381,7 @@ function createUpload(req, res, next) {
     var s = req.upload.mpuSize;
     var c = req.upload.mpuCopies;
 
-    chooseSharks(req, s, c, function (err, sharks) {
+    chooseSharks(req, res, s, c, function (err, sharks) {
         if (err) {
             next(err);
         } else {
diff --git a/package.json b/package.json
index 4c9e8a7..bea8507 100644
--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
         "backoff": "2.3.0",
         "bunyan": "0.22.1",
         "bunyan-syslog": "0.2.2",
-        "cueball": "2.2.9",
+        "cueball": "2.3.0",
         "dashdash": "1.3.2",
         "deep-equal": "0.0.0",
         "dtrace-provider": "0.2.8",
